### DISCUSSION TITLE
error in GPS measurements

### DISCUSSION THREAD

**Discussion by myarmatov** [microsoft/AirSim#3391] [Support Q&A] [answered]
Hi guys!
I have tried to measure the **GPS geopoint** data each second in **Multirotor** simulation, but two of the three measurements (**latitude and longitude**) do not change at all even though Multirotor flies with huge speed to the various directions. Did anyone encounter this kind of issue before? Or can anyone please explain its reason?
Thanks in advance!

**Comment by ahmed-elsaharti** [comment, marked as answer]
Hi @myarmatov , I just tested this with the blocks binary and the GPS location seems to update correctly.
Which API call are you using to obtain GPS data?
![GPSAirsim](https://user-images.githubusercontent.com/49626509/108022974-3dce7080-6fe7-11eb-8e06-12cc3c261e3d.gif)

**Reply by myarmatov** [reply]
Hi Sir,

Thank you for your reply. I have used getGpsData() function to obtain the GPS data. In your test, **Altitude** is changing as it should, however, **Latitude** (47.64...) and **Longitude** (-122.14...) are almost the same throughout the whole test. Results of my tests were very similar, but I thought they were incorrect. 

If you think your results make sense, could you please explain their unit or why very small amount of their value change?

**Reply by ahmed-elsaharti** [reply]
@myarmatov, the unit of latitude and longitude is degrees. The minor change you're seeing in each reading (~0.0001 degrees) is equivalent to ~11 meters each. You'll find more details about this [here](https://en.wikipedia.org/wiki/Decimal_degrees) and [here](https://www.usna.edu/Users/oceano/pguth/md_help/html/approx_equivalents.htm)

**Reply by myarmatov** [reply]
@ahmed-elsaharti, oh I got it now. I also thought it might be something like that, but I could not find the reference. Thank you so much for your help. But altitude's unit is meters, right? 
One last question: Do you know if AirSim simulates the GPS error as same as the real world? Because GPS measurements in real world can represent an error of about 3 meters, I cannot use the data obtained from AirSim if it does not simulate the same measurement error.

**Reply by ahmed-elsaharti** [reply]
@myarmatov , Glad this helped ^_^
Also, yes, as far as I remember the altitude should be in meters.
Regarding the question, I'm not entirely sure but I think Airsim reports ground truth GPS data (I could be wrong though). But as far as I know real world GPS error could go as low as [1m](https://www.flyingtech.co.uk/electronics/holybro-h-rtk-f9p-high-precision-gps-compass-module) or even [0.5m](https://www.amazon.com/Radiolink-Pixhawk-Controller-Quadcopter-Helicopter/dp/B07QPKR2FV#:~:text=High%20Positioning%20Accuracy%3A%20about%20of,few%20seconds%20at%20open%20ground.) and with an RTK system you could even reach centimeter accuracy

**Reply by zimmy87** [reply]
@myarmatov , as per GPS error: part of the output from getGpsData() is the struct defined [here](https://github.com/microsoft/AirSim/blob/master/AirLib/include/sensors/gps/GpsBase.hpp#L78), which contains two fields, eph and epv, which correspond to GPS HDOP/VDOP or [dilution of precision](https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation)), which I think are basically measurements of the GPS error. However, as best I can tell, these values aren't used in calculating the coordinates returned in getGpsData(), and instead coordinates are simply taken from ground truth. If you wanted to simulate the error, I think you would need to extend AirLib\include\sensors\gps\GpsSimple.hpp or AirLib\include\sensors\gps\GpsBase.hpp to add an error to coordinates returned by ground truth.
